File: ts/test/parity.test.js
/* Copyright (c) 2026 tabnas, MIT License */
'use strict'

// Cross-runtime conformance, driven by the shared `test/spec/*.tsv` fixtures
// at the repo root (see ../../test/AGENTS.md).
//
// The fixture loader, the escape codec, the `ERROR:<code>` contract and the
// row loop all come from @tabnas/support, whose Go half `go/parity_test.go`
// uses to run the SAME files — so the two implementations cannot drift
// without one of them going red, and neither can the two loaders.
//
// What is left here is only what is specific to @tabnas/json: every row is
// also cross-checked against the platform JSON parser, since this package's
// contract IS plain JSON.

const { findSpecDir, makeRunner } = require('@tabnas/support')

const { parse, TabnasError } = require('../dist/json.js')

// `JSON.stringify` with ONE correction: it renders negative zero as `0`,
// losing a distinction the value contract keeps (tabnas/support#13, and
// ADR-15). Go's `json.Marshal` renders it `-0`, so a fixture cell compared
// as text could not previously hold `-0` at all — the row was the only
// input in the suite where this runner and the Go one disagreed about what
// the `expected` column even means.
//
// Everything else defers to `JSON.stringify`, and objects are walked in
// `Object.keys` order, which is insertion order for string keys — so this
// still pins KEY ORDER, which is why the comparison is textual.
function render(value) {
  if ('number' === typeof value && Object.is(value, -0)) return '-0'
  if (Array.isArray(value)) return '[' + value.map(render).join(',') + ']'
  if (null !== value && 'object' === typeof value) {
    return '{' + Object.keys(value)
      .map((k) => JSON.stringify(k) + ':' + render(value[k])).join(',') + '}'
  }
  return JSON.stringify(value)
}

makeRunner({
  parse: (input) => {
    const value = parse(input)
    const text = render(value)

    // The platform parser must accept it, and produce the same thing. A
    // throw here surfaces as a failed row, naming the fixture and line.
    // Rendered the same way, so the oracle is held to the same contract
    // rather than being allowed to lose the sign where we do not.
    const platform = render(JSON.parse(input))
    if (text !== platform) {
      throw new Error(
        `disagrees with the platform parser: ${text} != ${platform}`)
    }

    return text
  },

  // Compared as TEXT, not structurally: the expected column is written
  // with `JSON.stringify` semantics (as corrected by `render` above), and
  // comparing the rendering is what pins KEY ORDER as well as value. (Go
  // pins order separately, in TestSpecValidOrder — it compares by value
  // here.)
  parseExpected: (expected) => expected,

  // Two sanity checks the code comparison would not otherwise make. Each
  // answers a pseudo-code rather than throwing, so the failure reads as
  // "failed with code <what went wrong>, expected <the row's code>" and
  // names the row.
  errorCode: (err, row) => {
    if (!(err instanceof TabnasError)) {
      return `not-a-TabnasError(${err})`
    }
    try {
      JSON.parse(row.unesc(0))
      return 'the-platform-parser-accepted-it'
    }
    catch {
      // Rejected by both, as it must be.
    }
    return err.code
  },
})
  // `findSpecDir` walks up from this file to the repo root's `test/spec`,
  // so moving the suite does not mean recounting `..` hops. `dir` then
  // auto-discovers every fixture in it, so adding a .tsv runs it in both
  // runtimes without touching either runner.
  .dir(findSpecDir(__dirname))

File: go/parity_test.go
// Copyright (c) 2026 tabnas, MIT License

package tabnasjson

// parity_test.go — cross-runtime conformance, driven by the shared
// `test/spec/*.tsv` fixtures at the repo root (see ../test/AGENTS.md).
//
// The fixture loader, the escape codec, the ERROR:<code> contract and the
// row loop all come from github.com/tabnas/support/go, whose TypeScript
// half ts/test/parity.test.js uses to run the SAME files — so the two
// implementations cannot drift without one of them going red, and neither
// can the two loaders.
//
// What is left here is only what is specific to @tabnas/json: every row is
// also cross-checked against encoding/json, since this package's contract
// IS plain JSON.

import (
	stdjson "encoding/json"
	"fmt"
	"testing"

	tabnas "github.com/tabnas/parser/go"
	support "github.com/tabnas/support/go"
)

// TestSpec runs every fixture in the spec directory. FindSpecDir walks up
// from the package directory, and Dir discovers the files by listing, so
// adding a .tsv runs it in both runtimes without touching either runner.
func TestSpec(t *testing.T) {
	dir, err := support.FindSpecDir("")
	if err != nil {
		t.Fatal(err)
	}

	support.Runner{
		Parse: func(input string) (any, error) {
			got, err := Parse(input)
			if err != nil {
				return nil, err
			}

			// The standard library must accept it, and produce the same
			// value. Returning the error surfaces as a failed row, naming
			// the fixture and line.
			var want any
			if err := stdjson.Unmarshal([]byte(input), &want); err != nil {
				return nil, fmt.Errorf(
					"encoding/json rejected a valid fixture: %w", err)
			}
			if !support.EqualValueWith(got, want, deorder) {
				return nil, fmt.Errorf(
					"disagrees with encoding/json: %s != %s",
					support.FormatValue(deorder(got)), support.FormatValue(want))
			}

			return got, nil
		},

		// Parsed objects are insertion-ordered *tabnas.OrderedMap; deorder
		// unwraps them so the comparison is by value. Key order is
		// separately covered by TestSpecValidOrder — TypeScript pins it in
		// this suite instead, by comparing the rendered text.
		Normalize: deorder,

		// Two sanity checks the code comparison would not otherwise make.
		// Each answers a pseudo-code rather than failing here, so the
		// failure reads as "failed with code <what went wrong>, expected
		// <the row's code>" and names the row.
		ErrorCode: func(err error) string {
			je, ok := err.(*tabnas.TabnasError)
			if !ok {
				return fmt.Sprintf("not-a-TabnasError(%T)", err)
			}
			return je.Code
		},
	}.Dir(t, dir)
}

// TestSpecRejectedByBoth pins the other half of the plain-JSON contract:
// every input a fixture says must FAIL must be rejected by encoding/json
// too. It is a separate walk rather than part of the runner's error branch
// because the runner hands the hook an error, not the input that caused it.
func TestSpecRejectedByBoth(t *testing.T) {
	dir, err := support.FindSpecDir("")
	if err != nil {
		t.Fatal(err)
	}

	specs, err := support.LoadSpecDir(dir, nil)
	if err != nil {
		t.Fatal(err)
	}

	checked := 0
	for _, spec := range specs {
		for _, row := range spec.Rows {
			if !support.IsErrorExpect(row.Col(1)) {
				continue
			}
			checked++

			input := row.Unesc(0)
			var sink any
			if nil == stdjson.Unmarshal([]byte(input), &sink) {
				t.Errorf("%s: encoding/json accepted %q, which the fixture "+
					"says must be rejected", row.Where(), input)
			}
		}
	}

	// A cross-check that walked no rows would be green having asserted
	// nothing, which is the failure mode this whole suite guards against.
	if 0 == checked {
		t.Fatal("no ERROR rows found — the cross-check asserted nothing")
	}
}

// TypeScript is canonical, and it reports this grammar's rules in the order
// they are declared: val, map, list, pair, elem. A Go map has no order, so
// without GrammarSpec.RuleOrder the engine falls back to sorted names and
// RuleNames answers [elem list map pair val] instead. That is not cosmetic:
// railroad's extracted Go model, and anything else built on RuleNames, renders
// in whatever order this returns.
//
// Hard-coded rather than derived, deliberately. Reading the order back out of
// the same literal that sets it would assert nothing; this is a copy of what
// ts/src/json.ts declares, so it fails if either side moves.
func TestRuleOrderMatchesTypeScriptDeclarationOrder(t *testing.T) {
	want := []string{"val", "map", "list", "pair", "elem"}
	got := Make().RuleNames()

	if len(got) != len(want) {
		t.Fatalf("RuleNames() = %v, want %v", got, want)
	}
	for i := range want {
		if got[i] != want[i] {
			t.Fatalf("RuleNames() = %v, want %v (differs at %d)", got, want, i)
		}
	}
}
